src/api/uploads/UploadsReachability.js

```diff
@@ -0,0 +1,194 @@
+        if (!localStoreResults) {
+            return;
+        }
+
+        Object.keys(localStoreResults).forEach(hostUrl => {
+            const result = localStoreResults[hostUrl];
+            if (this.isCachedHostValid(result)) {
+                this.cachedResults[hostUrl] = result;
+            }
+        });
+    }
+
+    /**
+     * Returns the host URLs that, according to the cached reachability test results, are unreachable
+     *
+     * @return {Array} The unreachable host URLs
+     */
+    getUnreachableHostsUrls() {
+        const unreachableHosts = [];
+        if (!this.cachedResults) {
+            // console.log('this.cachedResults is empty in getUnreachableHostsUrls');
+            return unreachableHosts;
+        }
+
+        Object.keys(this.cachedResults).forEach(hostUrl => {
+            const value = this.cachedResults[hostUrl];
+            if (this.isCachedHostValid(value) && !value.isReachable) {
+                // console.log(`${hostUrl} is not reachable`);
+                unreachableHosts.push(hostUrl);
+            }
+        });
+
+        return unreachableHosts;
+    }
+
+    /**
+     * Determines whether the given host is reachable by either making a test request to the uplaodHost
+     * or returning the result of the last reachability test it did
+     *
+     * @param {string} uploadHost - The uplaod host URL that will be stored in the cached test result and returned in
+     * getUnreachableHostsUrls() if test fails (this is usually a prefix of the uplaodUrl)
+     * @return {Promise<boolean>} Promise that resolved to true if the host is reachable, false if it is not
+     */
+    async isReachable(uploadHost: string) {
+        // The default upload host should always reachable
+        if (uploadHost === `${DEFAULT_HOSTNAME_UPLOAD}/`) {
+            return true;
+        }
+
+        const cachedResult = this.getCachedResult(uploadHost);
+        if (cachedResult) {
+            return cachedResult.isReachable;
+        }
+
+        const isHostReachable = await this.makeReachabilityRequest(uploadHost);
+        this.updateCachedResult(uploadHost, isHostReachable);
+        return isHostReachable;
+    }
+
+    /**
+     * Determines if the given hostUrl is reachable by making a test upload request to it.
+     * Does not read or modify any cached results.
+     *
+     * @param {string} hostUrl - The uplaod host urlto make a test request against
+     * @return {Promise<boolean>}
+     */
+    async makeReachabilityRequest(hostUrl: string) {
+        const url = `${hostUrl}html5?reachability_test=run`;
+        const headers: StringMap = {
+            [HEADER_CONTENT_TYPE]: 'application/x-www-form-urlencoded; charset=UTF-8',
+            'X-File-Name': 'reachability_pseudofile.txt',
+            'X-File-Size': '1234',
+        };
+        const data = 'reachability_file=test_file_data';
+
+        try {
+            await axios.post(url, data, { headers });
```

Comment:
Would this fail faster if we first check that a GET request succeeds before we attempt a POST?